File: src/syntax/latex/parser.ts
import { TokenBuffer } from '../token';
import {
  LatexCommandSyntax,
  LatexDocumentSyntax,
  LatexGroupSyntax,
  LatexSyntaxNode,
  LatexTextSyntax,
} from './ast';
import { LatexLexer, LatexToken, LatexTokenKind } from './lexer';

enum LatexScope {
  Document,
  Group,
  Options,
}

class LatexParser {
  constructor(private readonly tokens: TokenBuffer<LatexToken>) {}

  public document(): LatexDocumentSyntax {
    const children = this.content(LatexScope.Document);
    return new LatexDocumentSyntax(children);
  }

  private content(scope: LatexScope): LatexSyntaxNode[] {
    const children: LatexSyntaxNode[] = [];
    while (this.tokens.available) {
      switch (this.tokens.peek()!.kind) {
        case LatexTokenKind.Word:
        case LatexTokenKind.Math:
        case LatexTokenKind.BeginOptions:
          children.push(this.text(scope));
          break;
        case LatexTokenKind.Command:
          children.push(this.command());
          break;
        case LatexTokenKind.BeginGroup:
          children.push(this.group(LatexScope.Group));
          break;
        case LatexTokenKind.EndGroup:
          if (scope === LatexScope.Document) {
            this.tokens.next();
          } else {
            return children;
          }
          break;
        case LatexTokenKind.EndOptions:
          if (scope === LatexScope.Options) {
            return children;
          } else {
            children.push(this.text(scope));
          }
          break;
      }
    }
    return children;
  }

  private group(scope: LatexScope): LatexGroupSyntax {
    const left = this.tokens.next();
    const children = this.content(scope);
    const endKind =
      scope === LatexScope.Group
        ? LatexTokenKind.EndGroup
        : LatexTokenKind.EndOptions;

    const right = this.nextOfKind(endKind) ? this.tokens.next() : undefined;
    return new LatexGroupSyntax(left, children, right);
  }

  private command(): LatexCommandSyntax {
    const name = this.tokens.next();
    const options = this.nextOfKind(LatexTokenKind.BeginOptions)
      ? this.group(LatexScope.Options)
      : undefined;

    const args: LatexGroupSyntax[] = [];
    while (this.nextOfKind(LatexTokenKind.BeginGroup)) {
      args.push(this.group(LatexScope.Group));
    }

    return new LatexCommandSyntax(name, options, args);
  }

  private text(scope: LatexScope): LatexTextSyntax {
    const words: LatexToken[] = [];
    while (this.tokens.available) {
      const kind = this.tokens.peek()!.kind;
      const opts =
        kind === LatexTokenKind.EndOptions && scope !== LatexScope.Options;
      if (
        kind === LatexTokenKind.Word ||
        kind === LatexTokenKind.Math ||
        kind === LatexTokenKind.BeginOptions ||
        opts
      ) {
        words.push(this.tokens.next());
      } else {
        break;
      }
    }
    return new LatexTextSyntax(words);
  }

  private nextOfKind(kind: LatexTokenKind): boolean {
    const token = this.tokens.peek();
    return token !== undefined && token.kind === kind;
  }
}

export function parse(text: string): LatexDocumentSyntax {
  const lexer = new LatexLexer(text);
  const tokens = new TokenBuffer(lexer);
  const parser = new LatexParser(tokens);
  return parser.document();
}

File: crates/texlab_syntax/src/latex/parser.rs
use super::ast::*;
use std::iter::Peekable;
use std::sync::Arc;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum LatexScope {
    Root,
    Group,
    Options,
}

pub struct LatexParser<I: Iterator<Item = LatexToken>> {
    tokens: Peekable<I>,
}

impl<I: Iterator<Item = LatexToken>> LatexParser<I> {
    pub fn new(tokens: I) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn root(&mut self) -> LatexRoot {
        let children = self.content(LatexScope::Root);
        LatexRoot::new(children)
    }

    fn content(&mut self, scope: LatexScope) -> Vec<LatexContent> {
        let mut children = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            match token.kind {
                LatexTokenKind::Word | LatexTokenKind::BeginOptions => {
                    children.push(LatexContent::Text(self.text(scope)));
                }
                LatexTokenKind::Command => {
                    children.push(LatexContent::Command(self.command()));
                }
                LatexTokenKind::Comma => {
                    let node = LatexComma::new(self.tokens.next().unwrap());
                    children.push(LatexContent::Comma(Arc::new(node)));
                }
                LatexTokenKind::Math => {
                    children.push(LatexContent::Math(self.math()));
                }
                LatexTokenKind::BeginGroup => {
                    children.push(LatexContent::Group(self.group(LatexGroupKind::Group)));
                }
                LatexTokenKind::EndGroup => {
                    if scope == LatexScope::Root {
                        self.tokens.next();
                    } else {
                        return children;
                    }
                }
                LatexTokenKind::EndOptions => {
                    if scope == LatexScope::Options {
                        return children;
                    } else {
                        children.push(LatexContent::Text(self.text(scope)));
                    }
                }
            }
        }
        children
    }

    fn command(&mut self) -> Arc<LatexCommand> {
        let name = self.tokens.next().unwrap();

        let mut options = Vec::new();
        let mut args = Vec::new();
        while let Some(token) = self.tokens.peek() {
            match token.kind {
                LatexTokenKind::BeginGroup => {
                    args.push(self.group(LatexGroupKind::Group));
                }
                LatexTokenKind::BeginOptions => {
                    options.push(self.group(LatexGroupKind::Options));
                }
                _ => {
                    break;
                }
            }
        }
        Arc::new(LatexCommand::new(name, options, args))
    }

    fn group(&mut self, kind: LatexGroupKind) -> Arc<LatexGroup> {
        let left = self.tokens.next().unwrap();
        let scope = match kind {
            LatexGroupKind::Group => LatexScope::Group,
            LatexGroupKind::Options => LatexScope::Options,
        };
        let children = self.content(scope);
        let right_kind = match kind {
            LatexGroupKind::Group => LatexTokenKind::EndGroup,
            LatexGroupKind::Options => LatexTokenKind::EndOptions,
        };

        let right = if self.next_of_kind(right_kind) {
            self.tokens.next()
        } else {
            None
        };

        Arc::new(LatexGroup::new(left, children, right, kind))
    }

    fn text(&mut self, scope: LatexScope) -> Arc<LatexText> {
        let mut words = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            let kind = token.kind;
            let opts = kind == LatexTokenKind::EndOptions && scope != LatexScope::Options;
            if kind == LatexTokenKind::Word || kind == LatexTokenKind::BeginOptions || opts {
                words.push(self.tokens.next().unwrap());
            } else {
                break;
            }
        }
        Arc::new(LatexText::new(words))
    }

    fn math(&mut self) -> Arc<LatexMath> {
        let token = self.tokens.next().unwrap();
        Arc::new(LatexMath::new(token))
    }

    fn next_of_kind(&mut self, kind: LatexTokenKind) -> bool {
        if let Some(ref token) = self.tokens.peek() {
            token.kind == kind
        } else {
            false
        }
    }
}
